src/django_components/slots.py

```diff
@@ -1,41 +1,84 @@
 import difflib
 import json
-from copy import copy
-from typing import Dict, List, NamedTuple, Optional, Set, Type, Union
+import re
+from collections import deque
+from typing import Any, Dict, List, NamedTuple, Optional, Set, Tuple, Type
 
 from django.template import Context, Template
 from django.template.base import FilterExpression, Node, NodeList, Parser, TextNode
 from django.template.defaulttags import CommentNode
 from django.template.exceptions import TemplateSyntaxError
 from django.utils.safestring import SafeString, mark_safe
 
-from django_components.app_settings import SlotContextBehavior, app_settings
-from django_components.context import (
-    copy_forloop_context,
-    get_outer_root_context,
-    get_slot_component_association,
-    get_slot_fill,
-    set_slot_fill,
-)
+from django_components.app_settings import ContextBehavior, app_settings
+from django_components.context import _FILLED_SLOTS_CONTENT_CONTEXT_KEY, _ROOT_CTX_CONTEXT_KEY
 from django_components.logger import trace_msg
-from django_components.node import nodelist_has_content
+from django_components.node import nodelist_has_content, walk_nodelist, NodeTraverse
 from django_components.utils import gen_id
 
 DEFAULT_SLOT_KEY = "_DJANGO_COMPONENTS_DEFAULT_SLOT"
 
 # Type aliases
 
+SlotId = str
 SlotName = str
 AliasName = str
 
 
 class FillContent(NamedTuple):
-    """Data passed from component to slot to render that slot"""
+    """
+    This represents content set with the `{% fill %}` tag, e.g.:
+
+    ```django
+    {% component "my_comp" %}
+        {% fill "first_slot" %} <--- This
+            hi
+            {{ my_var }}
+            hello
+        {% endfill %}
+    {% endcomponent %}
+    ```
+    """
 
     nodes: NodeList
     alias: Optional[AliasName]
 
 
+class Slot(NamedTuple):
```

Comment:
Now we don't mutate the `SlotNode` and `FillNode` instances. Instead, we create instances of `Slot` and `SlotFill` to tell SlotNodes how they should be rendered.